File: x_to_telegram.py
#!/usr/bin/env python3
"""
Post new X (Twitter) posts from a user to a Telegram channel.

Requirements:
  pip install tweepy python-dotenv requests

Env (.env or environment):
  X_BEARER_TOKEN=xxxxxxxxxxxxxxxxxxxxxxxxxxxx
  TELEGRAM_BOT_TOKEN=1234567890:ABCDEF...
  TELEGRAM_CHAT_ID=@your_channel   # or -1001234567890

Typical GitHub Actions step:
  python x_to_telegram.py --once --max-per-run 50
"""

import argparse
import json
import os
import sys
import time
from datetime import timezone
from typing import List, Optional

import requests
import tweepy
from tweepy.errors import TooManyRequests
from dotenv import load_dotenv

# -------------------------
# Defaults (override via CLI)
# -------------------------
DEFAULT_USERNAME = "joecarlsonshow"
DEFAULT_STATE_FILE = "last_tweet_id.json"
DEFAULT_INTERVAL = 900          # seconds (15 min) when looping locally
DEFAULT_MAX_PER_RUN = 50        # safety cap (we request a single page anyway)

# -------------------------
# Utilities
# -------------------------
def log(msg: str):
    print(msg, flush=True)

def err(msg: str):
    print(msg, file=sys.stderr, flush=True)

def load_state(path: str) -> dict:
    if os.path.exists(path):
        try:
            with open(path, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception as e:
            err(f"[state] Could not read {path}: {e}")
    return {"last_id": None}

def save_state(path: str, last_id: Optional[str]):
    try:
        with open(path, "w", encoding="utf-8") as f:
            json.dump({"last_id": last_id}, f)
    except Exception as e:
        err(f"[state] Could not write {path}: {e}")

def chunk_telegram_message(text: str, chunk_size: int = 4096) -> List[str]:
    """Split long text into Telegram-safe chunks (max 4096 chars)."""
    if len(text) <= chunk_size:
        return [text]
    chunks = []
    start = 0
    while start < len(text):
        end = min(start + chunk_size, len(text))
        cut = text.rfind("\n", start, end)
        if cut == -1 or cut <= start:
            cut = end
        chunks.append(text[start:cut])
        start = cut
    return chunks

def extract_tweet_text(tweet: tweepy.Tweet) -> str:
    """Return the full tweet text, including Blue long-form tweets."""
    note_tweet = getattr(tweet, "note_tweet", None)
    if isinstance(note_tweet, dict):
        text = note_tweet.get("text")
        if text:
            return text.strip()

    # Tweepy 4.14+ exposes raw data mapping on tweet.data
    data = getattr(tweet, "data", None)
    if isinstance(data, dict):
        note = data.get("note_tweet")
        if isinstance(note, dict):
            text = note.get("text")
            if text:
                return text.strip()

    return (getattr(tweet, "text", "") or "").strip()


def build_message(username: str, tweet: tweepy.Tweet) -> str:
    created_utc = tweet.created_at.astimezone(timezone.utc).strftime("%Y-%m-%d %H:%M UTC") if tweet.created_at else "unknown time"
    tweet_url = f"https://x.com/{username}/status/{tweet.id}"
    text = extract_tweet_text(tweet)
    # Only include the tweet text and a single link to the tweet itself.
    # Previously we appended expanded URLs (e.g. pic.twitter.com links) which
    # resulted in multiple links being posted for tweets with photos/videos.
    lines = [
        text,
        "",
        tweet_url,
    ]

    return "\n".join(lines).strip()

def send_to_telegram(bot_token: str, chat_id: str, message: str, disable_preview: bool = False, dry_run: bool = False):
    if dry_run:
        log(f"[dry-run] Telegram message:\n{message}\n{'-'*40}")
        return

    api = f"https://api.telegram.org/bot{bot_token}/sendMessage"
    for part in chunk_telegram_message(message, 4096):
        payload = {
            "chat_id": chat_id,
            "text": part,
            "disable_web_page_preview": disable_preview,
            "parse_mode": "HTML",  # safe default; we don't rely on formatting
        }
        r = requests.post(api, json=payload, timeout=20)
        if r.status_code != 200:
            err(f"[telegram] HTTP {r.status_code} {r.text}")
            r.raise_for_status()

# -------------------------
# X (Twitter) client helpers
# -------------------------
def x_client_from_env() -> tweepy.Client:
    bearer = os.getenv("X_BEARER_TOKEN")
    if not bearer:
        raise SystemExit("Missing X_BEARER_TOKEN. Set it in your environment or .env file.")
    # IMPORTANT: don't auto-sleep; fail fast on rate limit
    return tweepy.Client(bearer_token=bearer, wait_on_rate_limit=False)

def get_user_id(client: tweepy.Client, username: str) -> str:
    resp = client.get_user(username=username)
    if not resp or not resp.data:
        raise SystemExit(f"User @{username} not found.")
    return str(resp.data.id)

def fetch_new_tweets(
    client: tweepy.Client,
    user_id: str,
    since_id: Optional[str],
    include_retweets: bool,
    include_replies: bool,
    max_per_run: int,
):
    """
    Fetch a single page from the user's timeline (newest first).
    We sort ascending so we can post oldest -> newest.
    Only ONE API call per run to avoid hitting quotas.
    """
    exclude = []
    if not include_retweets:
        exclude.append("retweets")
    if not include_replies:
        exclude.append("replies")

    resp = client.get_users_tweets(
        id=user_id,
        since_id=since_id,
        max_results=min(100, max_per_run),
    tweet_fields=["created_at", "entities", "note_tweet"],
        exclude=exclude or None,
    )

    tweets = list(resp.data or [])
    tweets.sort(key=lambda t: int(t.id))  # oldest -> newest
    return tweets[:max_per_run]

# -------------------------
# Core run step
# -------------------------
def run_once(
    username: str,
    state_file: str,
    telegram_token: str,
    telegram_chat_id: str,
    include_retweets: bool,
    include_replies: bool,
    max_per_run: int,
    disable_preview: bool,
    dry_run: bool,
) -> Optional[str]:
    client = x_client_from_env()
    user_id = get_user_id(client, username)

    state = load_state(state_file)
    last_id = state.get("last_id")

    try:
        tweets = fetch_new_tweets(
            client=client,
            user_id=user_id,
            since_id=last_id,
            include_retweets=include_retweets,
            include_replies=include_replies,
            max_per_run=max_per_run,
        )
    except TooManyRequests:
        # Fail fast so CI runners don't sit idle for ~10 minutes
        log("[warning] X API rate limit reached — skipping this run.")
        return last_id

    if not tweets:
        log("[info] No new tweets.")
        return last_id

    for t in tweets:
        msg = build_message(username, t)
        send_to_telegram(telegram_token, telegram_chat_id, msg, disable_preview, dry_run)
        last_id = str(t.id)
        save_state(state_file, last_id)
        log(f"[posted] {t.id}")

    return last_id

# -------------------------
# Main / CLI
# -------------------------
def main():
    load_dotenv()

    parser = argparse.ArgumentParser(description="Post new X posts to Telegram channel.")
    parser.add_argument("--username", default=DEFAULT_USERNAME, help="X handle without @")
    parser.add_argument("--state-file", default=DEFAULT_STATE_FILE, help="Path to last_id state file (json)")
    parser.add_argument("--interval", type=int, default=DEFAULT_INTERVAL, help="Loop interval seconds (ignored with --once)")
    parser.add_argument("--once", action="store_true", help="Run a single cycle and exit")
    parser.add_argument("--include-retweets", action="store_true", help="Also post retweets")
    parser.add_argument("--include-replies", action="store_true", help="Also post replies")
    parser.add_argument("--max-per-run", type=int, default=DEFAULT_MAX_PER_RUN, help="Max tweets to post per run (safety cap)")
    parser.add_argument("--disable-preview", action="store_true", help="Disable Telegram link previews")
    parser.add_argument("--dry-run", action="store_true", help="Do everything except sending to Telegram")

    args = parser.parse_args()

    tg_token = os.getenv("TELEGRAM_BOT_TOKEN")
    tg_chat = os.getenv("TELEGRAM_CHAT_ID")

    if not tg_token or not tg_chat:
        raise SystemExit("Missing TELEGRAM_BOT_TOKEN or TELEGRAM_CHAT_ID. Set them in your environment or .env file.")

    if args.once:
        try:
            run_once(
                username=args.username,
                state_file=args.state_file,
                telegram_token=tg_token,
                telegram_chat_id=tg_chat,
                include_retweets=args.include_retweets,
                include_replies=args.include_replies,
                max_per_run=max(1, args.max_per_run),
                disable_preview=args.disable_preview,
                dry_run=args.dry_run,
            )
        except TooManyRequests:
            log("[warning] Rate limited at top-level — exiting early.")
            sys.exit(0)
        return

    # Looping (useful if you run it on your own machine; for Actions use --once)
    log(f"[start] Polling @{args.username} every {args.interval}s. Press Ctrl+C to stop.")
    while True:
        try:
            run_once(
                username=args.username,
                state_file=args.state_file,
                telegram_token=tg_token,
                telegram_chat_id=tg_chat,
                include_retweets=args.include_retweets,
                include_replies=args.include_replies,
                max_per_run=max(1, args.max_per_run),
                disable_preview=args.disable_preview,
                dry_run=args.dry_run,
            )
        except TooManyRequests:
            log("[warning] Rate limited in loop — sleeping 300s then continuing.")
            time.sleep(300)
        except KeyboardInterrupt:
            log("\n[stop] Exiting by user request.")
            break
        except Exception as e:
            err(f"[error] {e}")
        time.sleep(max(5, args.interval))

if __name__ == "__main__":
    main()

File: x_to_telegram.js
#!/usr/bin/env node

/**
 * Post new X (Twitter) posts from a user to a Telegram channel.
 * Setup, environment variables, and usage: see README.md
 */

import { TwitterApi } from 'twitter-api-v2';
import { Bot } from 'grammy';
import yargs from 'yargs/yargs';
import { hideBin } from 'yargs/helpers';

// -------------------------
// Defaults
// -------------------------
const DEFAULT_MAX_PER_RUN = 50;
const TELEGRAM_MAX_MESSAGE_LENGTH = 4096;
const TELEGRAM_CAPTION_MAX_LENGTH = 1024;
const TELEGRAM_MEDIA_GROUP_MAX_SIZE = 10;
const X_API_MAX_RESULTS = 100;
const GIST_STATE_FILE = 'x-telegram-sync.json';

// -------------------------
// Utilities
// -------------------------
function getTimestamp() {
    return new Date().toISOString();
}

function log(msg) {
    console.log(`${getTimestamp()} ${msg}`);
}

function logError(msg) {
    console.error(`${getTimestamp()} ${msg}`);
}

function getGistCredentials() {
    const gistId = process.env.STATE_GIST_ID;
    const gistToken = process.env.GIST_TOKEN;
    if (!gistId || !gistToken) {
        throw new Error("Missing STATE_GIST_ID or GIST_TOKEN.");
    }
    return { gistId, gistToken };
}

function getGistUrl(gistId) {
    return `https://api.github.com/gists/${gistId}`;
}

function getGistHeaders(gistToken) {
    return {
        'Authorization': `Bearer ${gistToken}`,
        'Accept': 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
    };
}

async function fetchGist(gistId, gistToken) {
    const response = await fetch(getGistUrl(gistId), {
        headers: getGistHeaders(gistToken)
    });
    if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`GitHub API error: ${response.status} - ${errorText}`);
    }
    return response.json();
}

async function updateGist(gistId, gistToken, content) {
    const response = await fetch(getGistUrl(gistId), {
        method: 'PATCH',
        headers: { ...getGistHeaders(gistToken), 'Content-Type': 'application/json' },
        body: JSON.stringify({
            files: {
                [GIST_STATE_FILE]: { content: JSON.stringify(content) }
            }
        })
    });
    if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`GitHub API error: ${response.status} - ${errorText}`);
    }
}

async function loadState() {
    const { gistId, gistToken } = getGistCredentials();

    try {
        const gist = await fetchGist(gistId, gistToken);
        const content = gist.files[GIST_STATE_FILE]?.content;
        if (content) {
            const state = JSON.parse(content);
            log(`[state] Loaded from Gist: last_id=${state.last_id || 'none'}, user_id=${state.user_id || 'none'}`);
            return state;
        }
        log(`[state] Gist exists but ${GIST_STATE_FILE} not found, starting fresh`);
        return { last_id: null, user_id: null };
    } catch (e) {
        logError(`[state] CRITICAL: Could not load from Gist: ${e.message}`);
        throw e;
    }
}

async function saveState(lastId, userId) {
    const { gistId, gistToken } = getGistCredentials();

    try {
        await updateGist(gistId, gistToken, { last_id: lastId, user_id: userId });
        log(`[state] Saved to Gist: ${lastId}`);
    } catch (e) {
        logError(`[state] CRITICAL: Could not save to Gist: ${e.message}`);
        throw e;
    }
}

function chunkTelegramMessage(text, chunkSize = TELEGRAM_MAX_MESSAGE_LENGTH) {
    if (text.length <= chunkSize) return [text];

    const chunks = [];
    let start = 0;

    while (start < text.length) {
        const end = Math.min(start + chunkSize, text.length);
        let splitAt = text.lastIndexOf('\n', end);

        if (splitAt === -1 || splitAt <= start) {
            splitAt = end;
        }

        chunks.push(text.substring(start, splitAt));
        start = splitAt === end ? end : splitAt + 1;
    }

    return chunks;
}

function extractTweetText(tweet) {
    let text = (tweet.note_tweet?.text ?? tweet.text ?? "").trim();

    // Remove t.co URLs since the source link is already included
    const urls = tweet.entities?.urls || [];
    for (const url of urls) {
        if (url.url) {
            text = text.replaceAll(url.url, '');
        }
    }

    // Normalize whitespace: preserve newlines, collapse spaces within lines
    text = text.split('\n').map(line => line.replace(/ {2,}/g, ' ').trim()).join('\n');
    // Collapse 3+ consecutive newlines into 2
    return text.replace(/\n{3,}/g, '\n\n').trim();
}

function buildMessage(username, tweet, hasVideoPreview = false) {
    const tweetUrl = `https://x.com/${username}/status/${tweet.id}`;
    const tweetText = extractTweetText(tweet);

    let message = tweetText;
    if (hasVideoPreview) {
        message += '\n\n🎬 Watch the full video via the Source link below 👇';
    }
    message += `\n\n🔗 Source:\n${tweetUrl}`;

    return message;
}

function extractMedia(tweet, mediaData) {
    const mediaKeys = tweet.attachments?.media_keys || [];
    if (mediaKeys.length === 0 || mediaData.length === 0) return [];

    // Create lookup map for O(1) access
    const mediaMap = new Map(mediaData.map(m => [m.media_key, m]));

    return mediaKeys.map(key => mediaMap.get(key)).filter(media => {
        if (!media) return false;
        // Photos have url, videos only have preview_image_url in free tier
        if (media.type === 'photo' && media.url) return true;
        if (media.type === 'video' && media.preview_image_url) return true;
        return false;
    }).map(media => ({
        url: media.type === 'video' ? media.preview_image_url : media.url,
        isVideoPreview: media.type === 'video'
    }));
}

function generateCaption(message) {
    const ellipsis = '...';
    return message.length <= TELEGRAM_CAPTION_MAX_LENGTH
        ? message
        : message.substring(0, TELEGRAM_CAPTION_MAX_LENGTH - ellipsis.length) + ellipsis;
}

async function sendTextMessageChunked(bot, chatId, message) {
    const chunks = chunkTelegramMessage(message);
    for (const part of chunks) {
        await bot.api.sendMessage(chatId, part);
    }
}

async function sendToTelegram({ bot, chatId, message, media = [] }) {
    if (media.length === 0) {
        await sendTextMessageChunked(bot, chatId, message);
        return;
    }

    // Send media with caption; fall back to text-only if the media send fails
    try {
        if (media.length === 1) {
            await sendSingleMedia(bot, chatId, message, media[0]);
        } else {
            await sendMediaGroup(bot, chatId, message, media);
        }
    } catch (e) {
        logError(`[telegram] Media send failed, falling back to text: ${e.message}`);
        await sendTextMessageChunked(bot, chatId, message);
    }
}

async function sendSingleMedia(bot, chatId, message, mediaItem) {
    const caption = generateCaption(message);
    // All media (including video previews) are sent as photos
    await bot.api.sendPhoto(chatId, mediaItem.url, { caption });
}

async function sendMediaGroup(bot, chatId, message, media) {
    const mediaGroup = media.slice(0, TELEGRAM_MEDIA_GROUP_MAX_SIZE).map((item, idx) => ({
        // Photos and video previews are both sent to Telegram as photos
        type: 'photo',
        media: item.url,
        caption: idx === 0 ? generateCaption(message) : undefined
    }));

    await bot.api.sendMediaGroup(chatId, mediaGroup);

    // If caption was too long or there are more than max media items, send text separately
    if (message.length > TELEGRAM_CAPTION_MAX_LENGTH || media.length > TELEGRAM_MEDIA_GROUP_MAX_SIZE) {
        await sendTextMessageChunked(bot, chatId, message);
    }
}

// -------------------------
// X (Twitter) client helpers
// -------------------------
function getClient() {
    const bearer = process.env.X_BEARER_TOKEN;
    if (!bearer) {
        throw new Error("Missing X_BEARER_TOKEN.");
    }
    return new TwitterApi(bearer);
}

async function getUserId(client, username) {
    const user = await client.v2.userByUsername(username);

    if (!user.data) {
        throw new Error(`User @${username} not found.`);
    }
    return user.data.id;
}

async function fetchNewTweets(client, userId, sinceId, maxPerRun) {
    const tweets = await client.v2.userTimeline(userId, {
        since_id: sinceId || undefined,
        max_results: Math.min(X_API_MAX_RESULTS, maxPerRun),
        'tweet.fields': ['note_tweet', 'attachments', 'entities'],
        'expansions': ['attachments.media_keys'],
        'media.fields': ['url', 'preview_image_url', 'type'],
        exclude: ['retweets', 'replies']
    });

    // tweets.data is the API response; .data within it is the array of tweet objects
    const data = tweets.data?.data || [];
    const includes = tweets.data?.includes || {};

    // Sort oldest -> newest (API returns newest first)
    // Tweet IDs are numeric strings that sort lexicographically
    data.sort((a, b) => a.id.localeCompare(b.id));

    return { tweets: data, media: includes.media || [] };
}

// -------------------------
// Core run
// -------------------------
async function getUserIdWithCache(client, username, state) {
    if (state.user_id) {
        return state.user_id;
    }

    log(`[info] Looking up user ID for @${username}`);
    const userId = await getUserId(client, username);

    // Cache the user_id immediately to avoid redundant lookups
    await saveState(state.last_id, userId);
    return userId;
}

async function processTweets(tweets, mediaData, { username, tgToken, chatId, userId }) {
    const bot = new Bot(tgToken);

    for (const tweet of tweets) {
        const media = extractMedia(tweet, mediaData);
        const hasVideoPreview = media.some(m => m.isVideoPreview);
        const msg = buildMessage(username, tweet, hasVideoPreview);

        await sendToTelegram({ bot, chatId, message: msg, media });
        await saveState(tweet.id, userId);

        const mediaInfo = media.length > 0
            ? ` (with ${media.length} media: ${media.map(m => m.isVideoPreview ? 'video preview' : 'photo').join(', ')})`
            : '';

        log(`[posted] ${tweet.id}${mediaInfo}`);
    }
}

async function run(args) {
    const { username, maxPerRun } = args;
    const tgToken = process.env.TELEGRAM_BOT_TOKEN;
    const tgChat = process.env.TELEGRAM_CHAT_ID;

    if (!tgToken || !tgChat) {
        throw new Error('Missing TELEGRAM_BOT_TOKEN or TELEGRAM_CHAT_ID.');
    }

    const client = getClient();
    const state = await loadState();

    try {
        const userId = await getUserIdWithCache(client, username, state);
        const { tweets, media } = await fetchNewTweets(client, userId, state.last_id, maxPerRun);

        if (tweets.length === 0) {
            log('[info] No new tweets.');
            return;
        }

        log(`[info] Found ${tweets.length} new tweet(s)`);

        await processTweets(tweets, media, { username, tgToken, chatId: tgChat, userId });

        log(`[done] Posted ${tweets.length} tweet(s)`);
    } catch (e) {
        const statusCode = e.code ?? e.statusCode ?? e.response?.status ?? e.data?.status;

        if (statusCode === 429) {
            log('[warning] X API rate limit reached — skipping this run.');
            return;
        }

        if (statusCode === 402) {
            log('[warning] X API returned 402 Payment Required — the current API plan does not include this endpoint. Skipping this run.');
            return;
        }

        // Transient server-side errors (X API outage/hiccup). Skip this run;
        // the next scheduled run will pick up from the same last_id.
        if (statusCode >= 500 && statusCode <= 599) {
            log(`[warning] X API returned ${statusCode} — transient server error. Skipping this run.`);
            return;
        }

        throw e;
    }
}

// -------------------------
// Main / CLI
// -------------------------
async function main() {
    const argv = await yargs(hideBin(process.argv))
        .parserConfiguration({ 'camel-case-expansion': true })
        .option('username', { default: process.env.X_USERNAME, describe: 'X handle without @', demandOption: !process.env.X_USERNAME })
        .option('max-per-run', { default: DEFAULT_MAX_PER_RUN, type: 'number', describe: 'Max tweets to post per run' })
        .help()
        .argv;

    log(`[start] Checking @${argv.username}`);

    try {
        await run(argv);
    } catch (e) {
        logError(`[error] ${e.message}`);
        process.exit(1);
    }
}

if (process.argv[1] === new URL(import.meta.url).pathname) {
    main();
}
